c7n/resources/ami.py

```diff
@@ -74,3 +77,43 @@ class ImageAgeFilter(AgeFilter):
         'image-age',
         op={'type': 'string', 'enum': OPERATORS.keys()},
         days={'type': 'number', 'minimum': 0})
+
+
+@filters.register('unused')
+class ImageUnusedFilter(Filter):
+
+    schema = type_schema('unused', value={'type': 'boolean'})
+
+    def _pull_asg_images(self, results):
+        asg_manager = ASG(self.manager.ctx, {})
+        asgs = asg_manager.resources()
+        lcfgs = set(a['LaunchConfigurationName'] for a in asgs)
+
+        lcfg_manager = LaunchConfig(self.manager.ctx, {})
+        for lcfg in lcfg_manager.resources():
+            if lcfg['LaunchConfigurationName'] in lcfgs:
+                results.update(lcfg['ImageId'])
+        return results
+
+    def _pull_ec2_images(self):
+        ec2_manager = EC2(self.manager.ctx, {})
+        results = set()
+        results.update([i['ImageId'] for i in ec2_manager.resources()])
+        return results
+
+    def process(self, resources, event=None):
+        inuse_amis = self._pull_ec2_images()
+        inuse_amis = self._pull_asg_images(inuse_amis)
```

Comment:
rather than pass transient result variables as params, just keep the fetch images independent and union the results.

images = self._pull_ec2_images()
images = images.union(self._pull_asg_images())

or

self._pull_ec2_images().union(self._pull_asg_images())